### PR TITLE
feat: add sorting feature to summarylist

### PR DESCRIPTION
**Pull request by petter-a**
While setting the sortField property on a category in form builder,
the summarylist will sort the answers based on that field
Reach out to me and I will provide a walkthrough

### CODE REVIEW

**Comment by sebastianthulin** [comment]
Quality gate Quality gate: **_Passed_** 
 


            
 No issues Everything looks good
 [Details](https://app.embold.io/organization/gh/helsingborg-stad/repositories/fedfce3fd8695f238dea9cb00bd37ce7/pull-request/660)

**Comment by helsingborgsstad** [comment]
Task linked: [CU-221wakq Sortera svaren i summarylist per kategori i appen](https://app.clickup.com/t/221wakq)

File: source/components/molecules/GroupedList/GroupedList.tsx [hunk 1]
```diff
@@ -66,11 +80,18 @@ const GroupedList: React.FC<Props> = ({
         property: LayoutAnimation.Properties.opacity,
       },
     });
+    // Sort whenever the summarylist is closed for editing
+    // i.e the 'Close' button is tapped
```
Comment (MasterKenth) [inline source/components/molecules/GroupedList/GroupedList.tsx:84]: Don't think this comment is needed; it doesn't seem to add any information not presented as code.

File: source/components/molecules/GroupedList/GroupedList.tsx [hunk 2]
```diff
@@ -66,11 +80,18 @@ const GroupedList: React.FC<Props> = ({
...
+    if (editable && sortCallback) {
+      sortCallback(children.map((element) => element.props.item));
+    }
     setEditable(!editable);
   };
 
   categories.forEach((cat) => {
-    const catItems = children.filter((item) => item.props.category === cat.category);
+    const catItems = children.filter(
```
Comment (MasterKenth) [inline source/components/molecules/GroupedList/GroupedList.tsx:92]: Cat? 🐱. Prefer to not use abbreviations when possible.

File: source/components/organisms/SummaryList/SummaryList.tsx [hunk 3]
```diff
@@ -128,9 +129,30 @@ const SummaryList: React.FC<Props> = ({
...
+        const sortField = category?.sortField;
+        if (sortField && Array.isArray(answers[item.id])) {
+          const answerCopy = [...answers[item.id]].sort(
+            (valueA: Answer, valueB: Answer) => {
+              const a = Number(valueA[sortField]);
+              const b = Number(valueB[sortField]);
+              return a - b;
+            }
+          );
+          answers[item.id].splice(0);
```
Comment (MasterKenth) [inline source/components/organisms/SummaryList/SummaryList.tsx:149]:
```typescript
answers[item.id].splice(0);
answers[item.id].push(...answerCopy);
```

I'm not sure what this is supposed to do. `splice(0)` essentially empties the array, and then we push a single element to it? Wouldn't `answers[item.id] = answerCopy` do the same thing?

However on the larger scale, it's very cryptic that `sortAnswers` sorts the array in place. I would prefer if we stick to a functional approach here: `const newListItems = sortAnswers(listItems)`.

File: source/components/organisms/SummaryList/SummaryList.tsx [hunk 4]
```diff
@@ -128,9 +129,30 @@ const SummaryList: React.FC<Props> = ({
...
+  const sortAnswers = (listItems: SummaryListItem[]) => {
+    listItems.forEach((item) => {
+      if (ArrayType.includes(item.type)) {
+        const category = categories?.find(
+          (categoryItem) => item.category === categoryItem.category
+        );
+        const sortField = category?.sortField;
+        if (sortField && Array.isArray(answers[item.id])) {
+          const answerCopy = [...answers[item.id]].sort(
+            (valueA: Answer, valueB: Answer) => {
```
Comment (MasterKenth) [inline source/components/organisms/SummaryList/SummaryList.tsx:143]: I think the sorting logic itself is complex enough to warrant refactoring into a separate function. Something like:

```typescript
const answerCopy = [...answers[item.id]].sort((a, b) => compareAnswerBySortField(a, b, sortField));
```

File: source/components/organisms/SummaryList/SummaryList.tsx [hunk 5]
```diff
@@ -128,9 +129,30 @@ const SummaryList: React.FC<Props> = ({
   help,
   editable,
 }) => {
-  // For development: show an error popup when something is configured wrong.
-  const showNotification = useNotification();
+  const [isModified, setModified] = useState(false);
```
Comment (MasterKenth) [inline source/components/organisms/SummaryList/SummaryList.tsx:132]: `isModified` doesn't seem to be used?

File: source/components/organisms/SummaryList/SummaryList.tsx [hunk 6]
```diff
@@ -251,24 +273,31 @@ const SummaryList: React.FC<Props> = ({
...
         answers: answers[item.id] ?? "",
       });
     } else {
-      map.get(item.id)?.items.push(item);
+      itemIdmap.get(item.id)?.items.push(item);
     }
   });
 
+  useEffect(() => {
+    sortAnswers(itemsWithAnswers);
```
Comment (MasterKenth) [inline source/components/organisms/SummaryList/SummaryList.tsx:291]: Why is this wrapped in an `useEffect`? `itemsWithAnswers` is not stateful data (although it probably should be).

**Review by MasterKenth — CHANGES_REQUESTED** [review]
Overall I'm kind of confused as to both how this works and how it's implemented. Let's discuss face-to-face @petter-a 👌.

File: source/components/molecules/GroupedList/GroupedList.tsx (hunk 1 above)
Comment (petter-a) [reply source/components/molecules/GroupedList/GroupedList.tsx:84]: Fixed

File: source/components/molecules/GroupedList/GroupedList.tsx (hunk 2 above)
Comment (petter-a) [reply source/components/molecules/GroupedList/GroupedList.tsx:92]: Have not changed that but fixed it anyways

File: source/components/organisms/SummaryList/SummaryList.tsx (hunk 3 above)
Comment (petter-a) [reply source/components/organisms/SummaryList/SummaryList.tsx:149]: Yes I know but at the same time re-assignment is a bad practise so then I have to:
          // eslint-disable-next-line no-param-reassign
          answers[item.id] = answerCopy;

There is a reason why it has to be sorted in place, the architecture is built upon direct editing of the answers. Lets talk about it

File: source/components/organisms/SummaryList/SummaryList.tsx (hunk 4 above)
Comment (petter-a) [reply source/components/organisms/SummaryList/SummaryList.tsx:143]: Fixed

File: source/components/molecules/GroupedList/GroupedList.tsx (hunk 2 above)
Comment (MasterKenth) [reply source/components/molecules/GroupedList/GroupedList.tsx:92]: Ah It was just formatting, but might as well cleanup 👍.

File: source/components/molecules/GroupedList/GroupedList.tsx [hunk 7]
```diff
@@ -30,7 +34,12 @@ interface Props {
   showEditButton?: boolean;
   startEditable?: boolean;
   help?: Help;
-  children: React.ReactElement<{ category: string; editable?: boolean }>[];
+  children: React.ReactElement<{
+    category: string;
+    editable?: boolean;
+    item: SummaryListItem;
+  }>[];
+  sortCallback?: (items: SummaryListItem[]) => void;
```
Comment (MasterKenth) [inline source/components/molecules/GroupedList/GroupedList.tsx:42]: Generalize to `onCloseCallback` as discussed.

File: source/components/organisms/SummaryList/SummaryList.tsx [hunk 8]
```diff
@@ -128,9 +134,23 @@ const SummaryList: React.FC<Props> = ({
   help,
   editable,
 }) => {
-  // For development: show an error popup when something is configured wrong.
-  const showNotification = useNotification();
+  const [isModified, setModified] = useState(false);
 
+  const sortAnswers = (listItems: SummaryListItem[]) => {
```
Comment (MasterKenth) [inline source/components/organisms/SummaryList/SummaryList.tsx:139]: Rename `listItems` so it doesn't collide with same variable one scope up.

**Review by MasterKenth — CHANGES_REQUESTED** [review]
👌

File: source/components/molecules/GroupedList/GroupedList.tsx (hunk 7 above)
Comment (petter-a) [reply source/components/molecules/GroupedList/GroupedList.tsx]: Done

File: source/components/organisms/SummaryList/SummaryList.tsx (hunk 8 above)
Comment (petter-a) [reply source/components/organisms/SummaryList/SummaryList.tsx]: Done

**Review by MasterKenth — APPROVED** [review]

**Review by karlknowit — APPROVED** [review]

**Pull request merged**